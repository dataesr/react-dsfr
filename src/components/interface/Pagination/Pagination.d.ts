import * as React from 'react';

export type PaginationAnchorAs = 'a' | any;

export interface PaginationButtonLabels {
  currentAria?: string;
  firstAria?: string;
  lastAria?: string;
  navigationAria?: string;
  nextAria?: string;
  nextLabel?: string;
  pageAria?: (...args: any[])=>any;
  prevAria?: string;
  prevLabel?: string;
}

export interface PaginationProps {
  anchorAs?: PaginationAnchorAs;
  buildURL?: any;
  buttonLabels?: PaginationButtonLabels;
  currentPage: number;
  onClick?: any;
  pageCount: number;
  surrendingPages?: number;
}

declare const Pagination: React.FC<PaginationProps>;

export default Pagination;
