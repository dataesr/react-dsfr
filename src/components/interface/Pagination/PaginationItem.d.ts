import * as React from 'react';

export type PaginationItemAnchorAs = 'a' | any;

export interface PaginationItemProps {
  anchorAs?: PaginationItemAnchorAs;
  aria?: string;
  buildURL?: (...args: any[])=>any;
  isActive?: boolean;
  onClick?: (...args: any[])=>any;
  page: number;
}

declare const PaginationItem: React.FC<PaginationItemProps>;

export default PaginationItem;
