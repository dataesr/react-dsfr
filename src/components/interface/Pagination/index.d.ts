import type Pagination from './Pagination';
import type PaginationItem from './PaginationItem';

export type { Pagination, PaginationItem };
