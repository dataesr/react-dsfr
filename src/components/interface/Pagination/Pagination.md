## Avec gestion de la page par URL

```jsx
<Pagination buildURL={(page) => `my-path/page-${page}`} currentPage={8} pageCount={15} />
```

## Avec gestion de la page par state react

```jsx
import { useState } from 'react';

const [page, setPage] = useState(8);
<Pagination onClick={setPage} currentPage={page} pageCount={15} />
```
