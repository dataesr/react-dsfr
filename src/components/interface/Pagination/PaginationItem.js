import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const PaginationItem = ({
  page,
  anchorAs,
  isActive,
  isVisibleOnMobile,
  url,
  aria,
}) => {
  const Tag = `${anchorAs}`;
  const href = (anchorAs === 'a') ? url : undefined;
  const to = (anchorAs !== 'a') ? url : undefined;
  const _className = classNames('rf-pagination__item', {
    'rf-pagination__item--from-md': !isVisibleOnMobile,
    'rf-pagination__item--active': isActive,
  });
  if (isActive) return <li className={_className} aria-current={aria}>{page}</li>;
  return (
    <li className={_className}>
      <Tag
        href={href}
        to={to}
        className="rf-pagination__link"
        aria-label={aria}
        title={aria}
      >
        {page}
      </Tag>
    </li>
  );
};

PaginationItem.propTypes = {
  page: PropTypes.number.isRequired,
  isActive: PropTypes.bool,
  isVisibleOnMobile: PropTypes.bool,
  url: PropTypes.string,
  aria: PropTypes.string,
  anchorAs: PropTypes.oneOf(['a', PropTypes.elementType]),
};

PaginationItem.defaultProps = {
  anchorAs: 'a',
  isActive: false,
  isVisibleOnMobile: false,
  url: null,
  aria: null,
};
export default PaginationItem;
