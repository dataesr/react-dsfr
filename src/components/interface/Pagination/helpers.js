import range from '../../../utils/range';

// Calculation helpers
export const getSurrendingLeft = (x1, x2) => {
  if (x1 === 1) return { pages: [], hasEllipsis: false };
  if (x2 === 0) { return { pages: [], hasEllipsis: (x1 > 2) }; }
  if (x1 < x2 + 1) return { pages: range(2, x1 - 1), hasEllipsis: false };
  return { pages: range(x1 - x2, x1 - 1), hasEllipsis: true };
};
export const getSurrendingRight = (x1, x2, x3) => {
  if (x1 === x3) return { pages: [], hasEllipsis: false };
  if (x2 === 0) { return { pages: [], hasEllipsis: (x1 < x3 - 1) }; }
  if (Math.abs(x1 - x3) < x2 + 1) return { pages: range(x1 + 1, x3 - 1), hasEllipsis: false };
  return { pages: range(x1 + 1, x1 + x2), hasEllipsis: true };
};
