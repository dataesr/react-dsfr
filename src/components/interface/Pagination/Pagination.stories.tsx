import { ComponentStory } from '@storybook/react';

import Pagination from './Pagination';
import PaginationItem from './PaginationItem';

// https://www.systeme-de-design.gouv.fr/elements-d-interface/composants/pagination/

export default {
  title: 'Composants/Pagination - Pagination',
  component: Pagination,
  subcomponents: { PaginationItem },
  parameters: {
    docs: {
      description: {
        component: 'La pagination permet à l\'utilisateur de naviguer entre les différentes pages d\'une liste d\'éléments.',
      },
    },
  },
};

const Template: ComponentStory<typeof Pagination> = (args) => <Pagination {...args} />;

export const PaginationSimple = Template.bind({});
PaginationSimple.args = {
  currentPage: 1,
  pageCount: 20,
};
PaginationSimple.storyName = 'Simple';

export const PaginationButtonLabels = Template.bind({});
PaginationButtonLabels.args = {
  currentPage: 1,
  buttonLabels: {
    currentAria: 'currentAria',
    firstAria: 'firstAria',
    lastAria: 'lastAria',
    navigationAria: 'navigationAria',
    nextAria: 'nextAria',
    nextLabel: 'nextLabel',
    pageAria: (page) => `Page ${page}`,
    prevAria: 'prevAria',
    prevLabel: 'prevLabel',
  },
  pageCount: 20,
};
PaginationButtonLabels.storyName = 'Labels des boutons';

export const PaginationClick = Template.bind({});
PaginationClick.args = {
  currentPage: 1,
  // eslint-disable-next-line no-console
  onClick: () => console.log('click'),
  pageCount: 20,
};
PaginationClick.storyName = 'Evènement clic';

export const PaginationAnchor = Template.bind({});
PaginationAnchor.args = {
  anchorAs: 'a',
  currentPage: 1,
  pageCount: 20,
};
PaginationAnchor.storyName = 'Ancre';

export const PaginationBuildURL = Template.bind({});
PaginationBuildURL.args = {
  buildURL: (index: any) => `this is an url ${index}`,
  currentPage: 1,
  pageCount: 20,
};
PaginationBuildURL.storyName = 'Construction d\'URL';

export const PaginationSurrendingPages = Template.bind({});
PaginationSurrendingPages.args = {
  currentPage: 7,
  pageCount: 20,
  surrendingPages: 3,
};
PaginationSurrendingPages.storyName = 'Nombres de pages affichées de chaque côté';
