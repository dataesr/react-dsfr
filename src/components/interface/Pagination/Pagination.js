import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import dataAttributes from '../../../utils/data-attributes';
import { getSurrendingLeft, getSurrendingRight } from './helpers';
import PaginationItem from './PaginationItem';

/*
* DSFR v1.3.1
*/
import '@gouvfr/dsfr/dist/component/pagination/pagination.css';

const Pagination = ({
  pageCount,
  currentPage,
  anchorAs,
  buildURL,
  onClick,
  buttonLabels,
  surrendingPages,
  ...remainingProps
}) => {
  const surrendingLeft = getSurrendingLeft(currentPage, surrendingPages);
  const surrendingRight = getSurrendingRight(currentPage, surrendingPages, pageCount);
  const HTMLTag = `${anchorAs}`;
  const {
    navigationAria, prevLabel, nextLabel, prevAria,
    nextAria, firstAria, lastAria, currentAria, pageAria,
  } = buttonLabels;

  const getItem = (down, index, aria, icon, label) => {
    const check = down ? currentPage > 1 : currentPage < pageCount;
    if (buildURL) {
      return (
        <HTMLTag
          href={check ? buildURL(index) : undefined}
          className={classNames({ [`fr-pagination__link--${icon}`]: aria }, 'fr-pagination__link')}
          aria-label={aria}
          title={aria}
        >
          {label}
        </HTMLTag>
      );
    }

    return (
      <button
        type="button"
        onClick={() => onClick(index)}
        disabled={!check}
        className={classNames({ [`fr-pagination__link--${icon}`]: aria }, 'fr-pagination__link')}
        aria-label={aria}
      >
        {label}
      </button>
    );
  };

  return (
    <nav
      className="fr-pagination"
      aria-label={navigationAria}
      role="navigation"
      {...dataAttributes.getAll(remainingProps)}
    >
      <ul className="fr-pagination__list">
        <li>
          {getItem(true, 1, firstAria, 'first')}
        </li>
        <li>
          {getItem(true, currentPage - 1, prevAria, 'prev', prevLabel)}
        </li>
        <PaginationItem
          isVisibleOnMobile
          page={1}
          anchorAs={anchorAs}
          isActive={currentPage === 1}
          buildURL={buildURL}
          onClick={onClick}
          aria={pageAria(1)}
        />
        { surrendingLeft.hasEllipsis && (
          <li>…</li>)}
        {surrendingLeft.pages.map((p) => (
          <PaginationItem
            key={p}
            page={p}
            anchorAs={anchorAs}
            buildURL={buildURL}
            onClick={onClick}
            aria={pageAria(p)}
          />
        ))}
        {((currentPage !== 1) && (currentPage !== pageCount)) && (
          <PaginationItem
            page={currentPage}
            isVisibleOnMobile
            isActive
            aria={currentAria}
          />
        )}
        {surrendingRight.pages.map((p) => (
          <PaginationItem
            key={p}
            page={p}
            anchorAs={anchorAs}
            buildURL={buildURL}
            onClick={onClick}
            aria={pageAria(p)}
          />
        ))}
        { surrendingRight.hasEllipsis && (
          <li>…</li>)}
        {(pageCount !== 1) && (
          <PaginationItem
            aria={pageAria(pageCount)}
            isVisibleOnMobile
            aria-current={(currentPage === pageCount && 'page') || undefined}
            isActive={currentPage === pageCount}
            page={pageCount}
            anchorAs={anchorAs}
            buildURL={buildURL}
            onClick={onClick}
          />
        )}
        <li>
          {getItem(false, currentPage + 1, nextAria, 'next', nextLabel)}
        </li>
        <li>
          {getItem(false, pageCount, lastAria, 'last')}
        </li>
      </ul>
    </nav>
  );
};

Pagination.propTypes = {
  currentPage: PropTypes.number.isRequired,
  pageCount: PropTypes.number.isRequired,
  buildURL: (props, propName, componentName) => {
    const { buildURL, onClick } = props;
    if ((!buildURL && !onClick) || (buildURL && onClick)) {
      return new Error(`You must specify only one of props 'buildURL' or 'onClick' in '${componentName}'.`);
    }
    if (buildURL) {
      PropTypes.checkPropTypes(
        {
          buildURL: PropTypes.func,
        },
        { buildURL },
        'prop',
        'Pagination',
      );
    }
    return null;
  },
  onClick: (props, propName, componentName) => {
    const { buildURL, onClick } = props;
    if ((!buildURL && !onClick) || (buildURL && onClick)) {
      return new Error(`You must specify only one of props 'buildURL' or 'onClick' in '${componentName}'.`);
    }
    if (onClick) {
      PropTypes.checkPropTypes(
        {
          onClick: PropTypes.func,
        },
        { onClick },
        'prop',
        'Pagination',
      );
    }
    return null;
  },
  anchorAs: PropTypes.oneOf(['a', PropTypes.elementType]),
  buttonLabels: PropTypes.shape({
    navigationAria: PropTypes.string,
    currentAria: PropTypes.string,
    pageAria: PropTypes.func,
    prevLabel: PropTypes.string,
    nextLabel: PropTypes.string,
    prevAria: PropTypes.string,
    nextAria: PropTypes.string,
    firstAria: PropTypes.string,
    lastAria: PropTypes.string,
  }),
  surrendingPages: PropTypes.number,
};

Pagination.defaultProps = {
  anchorAs: 'a',
  surrendingPages: 2,
  buildURL: undefined,
  onClick: undefined,
  buttonLabels: {
    navigationAria: 'Pagination navigation',
    currentAria: 'page',
    pageAria: (page) => `Page ${page}`,
    prevLabel: 'Précédente',
    nextLabel: 'Suivante',
    prevAria: 'Page précédente',
    nextAria: 'Page suivante',
    firstAria: 'Première page',
    lastAria: 'Dernière Page',
  },
};
export default Pagination;
