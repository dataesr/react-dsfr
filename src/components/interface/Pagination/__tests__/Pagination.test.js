import { render, screen, within } from '@testing-library/react';
import Pagination from '..';

describe('<Pagination />', () => {
  it('should render pagination properly', () => {
    render(
      <Pagination buildURL={(page) => `page${page}`} currentPage={8} pageCount={15} />,
    );
    const pagination = screen.getByTestId('pagination-nav');
    expect(pagination.className).toBe('rf-pagination');
    const pages = screen.getByRole('list');
    const { getAllByRole } = within(pages);
    const items = getAllByRole('listitem');
    expect(items.length).toBe(13);
    expect(pagination).toMatchSnapshot();
  });
});
