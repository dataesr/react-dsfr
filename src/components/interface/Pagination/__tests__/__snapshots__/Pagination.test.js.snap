// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render pagination properly 1`] = `
<nav
  aria-label="Pagination navigation"
  class="rf-pagination"
  data-testid="pagination-nav"
>
  <ul
    class="rf-pagination__list"
  >
    <li
      class="rf-pagination__item--first rf-pagination__item "
    >
      <a
        aria-label="Première page"
        class="rf-pagination__link"
        title="Première page"
      />
    </li>
    <li
      class="rf-pagination__item--prev rf-pagination__item "
    >
      <a
        aria-label="Page précédente"
        class="rf-pagination__link"
        title="Page précédente"
      >
        <span
          class="rf-pagination__label"
        >
          Précédente
        </span>
      </a>
    </li>
    <li
      class="rf-pagination__item"
    >
      <a
        aria-label="Page 1"
        class="rf-pagination__link"
        href="page1"
        title="Page 1"
      >
        1
      </a>
    </li>
    <li
      class="rf-pagination__item rf-pagination__item--from-md"
    >
      …
    </li>
    <li
      class="rf-pagination__item rf-pagination__item--from-md"
    >
      <a
        aria-label="Page 6"
        class="rf-pagination__link"
        href="page6"
        title="Page 6"
      >
        6
      </a>
    </li>
    <li
      class="rf-pagination__item rf-pagination__item--from-md"
    >
      <a
        aria-label="Page 7"
        class="rf-pagination__link"
        href="page7"
        title="Page 7"
      >
        7
      </a>
    </li>
    <li
      aria-current="page"
      class="rf-pagination__item rf-pagination__item--active"
    >
      8
    </li>
    <li
      class="rf-pagination__item rf-pagination__item--from-md"
    >
      <a
        aria-label="Page 9"
        class="rf-pagination__link"
        href="page9"
        title="Page 9"
      >
        9
      </a>
    </li>
    <li
      class="rf-pagination__item rf-pagination__item--from-md"
    >
      <a
        aria-label="Page 10"
        class="rf-pagination__link"
        href="page10"
        title="Page 10"
      >
        10
      </a>
    </li>
    <li
      class="rf-pagination__item rf-pagination__item--from-md"
    >
      …
    </li>
    <li
      class="rf-pagination__item"
    >
      <a
        aria-label="Page 15"
        class="rf-pagination__link"
        href="page15"
        title="Page 15"
      >
        15
      </a>
    </li>
    <li
      class="rf-pagination__item--next rf-pagination__item "
    >
      <a
        aria-label="Page suivante"
        class="rf-pagination__link"
        title="Page suivante"
      >
        <span
          class="rf-pagination__label"
        >
          Suivante
        </span>
      </a>
    </li>
    <li
      class="rf-pagination__item--last rf-pagination__item "
    >
      <a
        aria-label="Dernière Page"
        class="rf-pagination__link"
        title="Dernière Page"
      />
    </li>
  </ul>
</nav>
`;
